chess/ChessGameModel.swift

```diff
@@ -15,22 +16,88 @@ struct ChessGameModel {
     }
 
     private(set) var board: [[ChessPiece?]] = .init(repeating: .init(repeating: nil, count: 8), count: 8)
+    private let logger = Logger(subsystem: "com.khaletska.chess", category: "GameModel")
 
     mutating func createNewGameBoard(configuration: BoardConfiguration) {
         self.board = configuration.generateBoard()
     }
 
-    mutating func move(from: Coordinate, to: Coordinate) {
-        let piece = self.board[from.row][from.col]
-        self.board[to.row][to.col] = piece
-        self.board[from.row][from.col] = nil
+    mutating func pieceMoved(from source: Coordinate, to destination: Coordinate) throws {
+        guard let piece = self.board[source.row][source.col] else {
+            throw ChessGameModelError.missingPiece
+        }
+
+        if isPawnPromotionMove(piece: piece, to: destination) {
+            try self.promotePawn(from: source, to: destination)
+            self.logger.log("Pawn \(piece) moved from \(source) to \(destination) and promoted to \(ChessPiece(kind: .queen, color: piece.color))")
+        }
+        else {
+            movePiece(from: source, to: destination)
+            self.logger.log("Piece \(piece) moved from \(source) to \(destination)")
+        }
+    }
+
+    // MARK: - Actions -
+
+    private mutating func movePiece(from source: Coordinate, to destination: Coordinate) {
+        self.board[destination.row][destination.col] = self.board[source.row][source.col]
+        self.board[source.row][source.col] = nil
+    }
+
+    private mutating func promotePawn(from source: Coordinate, to destination: Coordinate) throws {
+        guard let piece = self.board[source.row][source.col] else {
+            throw ChessGameModelError.missingPiece
+        }
+
+        guard case .pawn = piece.kind else {
+            throw ChessGameModelError.invalidPromotion
+        }
+
+        guard (piece.color == .white && destination.row == 0) || (piece.color == .black && destination.row == 7) else {
+            throw ChessGameModelError.invalidPromotion
+        }
+
+        self.board[destination.row][destination.col] = ChessPiece(kind: .queen, color: piece.color)
```

Comment:
Does it make sense to promote here and then call movePiece?
Or name this method "promote and move"

Reply:
promotion for me is equal to move. So basically there will be 3(4) options on `move` intention:
- eating a piece
- promotion
- casting(but probably can be made manually)
- just move

if you want each of these actions ended up with `move` it probably makes sense to rewrite this PR a bit to follow this idea. Especially since `eating` should follow the same logic

Reply:
let's make move part of promotion method